scylla/src/transport/load_balancing/latency_aware.rs

```diff
@@ -0,0 +1,704 @@
+    /// a 2nd chance. More precisely, a node is excluded from query plans if both his calculated
+    /// average latency is [exclusion_threshold](Self::exclusion_threshold) times slower than
+    /// the fastest node average latency (at the time the query plan is computed) **and** his
+    /// calculated average latency has been updated since less than [retry_period](Self::retry_period).
+    /// Since penalised nodes will likely not see their latency updated, this is basically how long
+    /// the policy will exclude a node.
+    pub retry_period: Duration,
+
+    /// The update rate defines how often the minimum average latency is recomputed. While the
+    /// average latency score of each node is computed iteratively (updated each time a new latency
+    /// is collected), the minimum score needs to be recomputed from scratch every time, which is
+    /// slightly more costly. For this reason, the minimum is only re-calculated at the given fixed
+    /// rate and cached between re-calculation.
+    /// The default update rate if **100 milliseconds**, which should be appropriate for most
+    /// applications. In particular, note that while we want to avoid to recompute the minimum for
+    /// every query, that computation is not particularly intensive either and there is no reason to
+    /// use a very slow rate (more than second is probably unnecessarily slow for instance).
+    pub update_rate: Duration,
+
+    /// Penalising nodes is based on an average of their recently measured average latency.
+    /// This average is only meaningful if a minimum of measurements have been collected.
+    /// This is what this option controls. If fewer than [minimum_measurements](Self::minimum_measurements)
+    /// data points have been collected for a given host, the policy will never penalise that host.
+    /// Note that the number of collected measurements for a given host is reset if the node
+    /// is restarted.
+    /// The default for this option is **50**.
+    pub minimum_measurements: usize,
+
+    /// Last minimum average latency that was noted among the nodes. It is updated every
+    /// [update_rate](Self::update_rate).
+    last_min_latency: Arc<AtomicDuration>,
+
+    child_policy: Box<dyn ChildLoadBalancingPolicy>,
+
+    nodes_tx: single_value_channel::Updater<Option<Vec<Arc<Node>>>>,
+    ask_for_nodes: Arc<AtomicBool>,
+
+    _updater_handle: RemoteHandle<()>,
+}
+
+impl LatencyAwarePolicy {
+    pub fn new(
+        exclusion_threshold: f64,
+        retry_period: Duration,
+        update_rate: Duration,
+        minimum_measurements: usize,
+        child_policy: Box<dyn ChildLoadBalancingPolicy>,
+    ) -> Self {
+        let min_latency = Arc::new(AtomicDuration::new());
+        let mut update_scheduler = tokio::time::interval(update_rate);
+        let ask_for_nodes = Arc::new(AtomicBool::new(true));
+        let (mut nodes_rx, nodes_tx) = single_value_channel::channel::<Vec<Arc<Node>>>();
```

Comment:
The single value channel has API that is distinct from bare `Arc<Mutex<Option<T>>>` in that it clearly separates writing and reading sides. Of course we can refrain from adding the dependency, but its purpose is valid here.

Reply:
After all, the whole purpose of using types is reducing the possible actions to performs on data to those which are purpose- and meaning- -ful

Reply:
I understand, but IMO it's not worth it. The usage of `nodes` is simple enough here that typing `nodes.lock().unwrap().clone()` instead of `nodes.latest()` doesn't obscure the algorithm too much, and it's not hard to track who is the writer and who is the reader. On the other hand, seeing a non-standard crate like this makes you wonder how it works and you have to learn a new abstraction.